zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -1304,4 +1304,70 @@ pub(crate) mod tests {
 
         assert_eq!(spendable.len(), 1);
     }
+
+    #[test]
+    fn checkpoint_gaps() {
+        let mut st = TestBuilder::new()
+            .with_block_cache()
+            .with_test_account(AccountBirthday::from_sapling_activation)
+            .build();
+
+        let (account, usk, birthday) = st.test_account().unwrap();
+        let dfvk = st.test_account_sapling().unwrap();
+
+        // Generate a block with funds belonging to our wallet.
+        st.generate_next_block(
+            &dfvk,
+            AddressType::DefaultExternal,
+            Amount::const_from_i64(500000),
+        );
+        st.scan_cached_blocks(birthday.height(), 1);
+
+        // Create a gap of 10 blocks having no shielded outputs, then add a block that doesn't
+        // belong to us so that we can get a checkpoint in the tree.
+        let not_our_key = ExtendedSpendingKey::master(&[]).to_diversifiable_full_viewing_key();
+        let not_our_value = Amount::const_from_i64(10000);
+        st.generate_block_at(
+            birthday.height() + 10,
+            BlockHash([0; 32]),
+            &not_our_key,
+            AddressType::DefaultExternal,
+            not_our_value,
+            st.latest_cached_block().unwrap().2,
+        );
+
+        // Scan the block
+        st.scan_cached_blocks(birthday.height() + 10, 1);
+
+        // Fake that everything has been scanned
+        st.wallet()
+            .conn
+            .execute_batch("UPDATE scan_queue SET priority = 10")
+            .unwrap();
+
+        // Verify that our note is considered spendable
+        let spendable = select_spendable_sapling_notes(
+            &st.wallet().conn,
+            account,
+            Amount::const_from_i64(300000),
+            birthday.height() + 5,
+            &[],
+        )
+        .unwrap();
+        assert_eq!(spendable.len(), 1);
+
+        // Attempt to spend the note with 5 confirmations
+        let to = not_our_key.default_address().1.into();
+        assert_matches!(
```

Comment:
In #908 I change tests like this to more thoroughly check that the spend succeeds (e.g. see https://github.com/zcash/librustzcash/pull/908/files#diff-9cea07a2b90189d927a2737a964494dfea6ab1bc5a51342e481210b93586716eR812-R835 ). I'd really like to get that PR in soon, so that we don't keep adding tests in the less thorough style.